### ISSUE TITLE
jinja2.exceptions.UndefinedError: 'context' is undefined

### ISSUE THREAD

**Issue by HADB** [python-semantic-release/python-semantic-release#984]
## The problem

jinja2.exceptions.UndefinedError: 'context' is undefined while using context in `.release_notes.md.j2`.

## Logs

Please provide debug logs for the command you are using with the `-vv` flag, e.g.

```shell
Traceback (most recent call last):
  File "/psr/.venv/bin/semantic-release", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/psr/.venv/lib/python3.12/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/psr/.venv/lib/python3.12/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/psr/.venv/lib/python3.12/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/psr/.venv/lib/python3.12/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/psr/.venv/lib/python3.12/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/psr/.venv/lib/python3.12/site-packages/click/decorators.py", line 45, in new_func
    return f(get_current_context().obj, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/psr/.venv/lib/python3.12/site-packages/semantic_release/cli/commands/version.py", line 701, in version
    release_notes = generate_release_notes(
                    ^^^^^^^^^^^^^^^^^^^^^^^
  File "/psr/.venv/lib/python3.12/site-packages/semantic_release/cli/changelog_writer.py", line 166, in generate_release_notes
    return render_release_notes(
           ^^^^^^^^^^^^^^^^^^^^^
  File "/psr/.venv/lib/python3.12/site-packages/semantic_release/cli/changelog_writer.py", line [57](https://github.com/HADB/python-semantic-release-test/actions/runs/10017951605/job/27692745834#step:4:58), in render_release_notes
    return template.render().rstrip()
           ^^^^^^^^^^^^^^^^^
  File "/psr/.venv/lib/python3.12/site-packages/jinja2/environment.py", line 1304, in render
    self.environment.handle_exception()
  File "/psr/.venv/lib/python3.12/site-packages/jinja2/environment.py", line 939, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "<template>", line 14, in top-level template code
  File "/psr/.venv/lib/python3.12/site-packages/jinja2/sandbox.py", line 327, in getattr
    value = getattr(obj, attribute)
            ^^^^^^^^^^^^^^^^^^^^^^^
jinja2.exceptions.UndefinedError: 'context' is undefined
```

**Comment by codejedi365** [comment]
The context variable is not defined for the release notes template unfortunately. It is only release and version globals directly.

I have recently uncovered that they have not been updated in line with the changelog environment but I have put effort towards improving it yet

**Comment by HADB** [comment]
Thanks for replaying.
If `context` variable is not usable in release notes template which I could use several months ago, what's the best practice to generate detailed changes compared with previous version like this: Detailed Changes: https://github.com/python-semantic-release/python-semantic-release/compare/v9.8.5...v9.8.6

**Comment by codejedi365** [comment]
> ... which I could use several months ago

Which version were you using? I have been refactoring changelog stuff but I don't remember context ever being available in release notes templates.

> what's the best practice to generate detailed changes compared with previous version

Unfortunately, it is not possible currently until I add the previous version in as an environment variable to the render context. If you are wondering about this projects release notes, I'm modifying it manually after the fact. My intention is to build a template for it soon and sus out the bugs & limitations here as you have illuminated.

**Comment by maxrake** [comment]
I get this same error, when using a custom `templates/.release_notes.md.j2` template:

```
{% set releases = context.history.released.items() | list %}
{% set prev_version = "v{}".format(releases[1][0]) %}
{% include ".changes.j2" %}
**Full Changelog**: {{ prev_version | compare_url("v{}".format(version)) }}
```

The last version of PSR I used to successfully create a release with this template was v9.8.3

**Comment by nfelt14** [comment]
I also had to downgrade to get it working. Will this eventually get fixed?

**Comment by codejedi365** [comment]
Yes, I will revert the change soon. Although this never was intentional to be exposed. What portion of the context are you using? The previous version extraction as well?

**Comment by nfelt14** [comment]
> @nfelt14, Yes, I will revert the change soon. Although this never was intentional to be exposed. What portion of the context are you using? The previous version extraction as well?

I am using `(context.history.released.items()|list)[0]`

**Comment by codejedi365** [comment]
@nfelt14, that looks like you are just pulling the current version number (i.e. the latest). In your case you don't need context at all. 

For release notes (not changelog generation), we provide both the release and version global variable. You can see an example of this in the built in default [`release_notes.j2`]() file.

**Comment by nfelt14** [comment]
> @nfelt14, that looks like you are just pulling the current version number (i.e. the latest). In your case you don't need context at all.
> 
> For release notes (not changelog generation), we provide both the release and version global variable. You can see an example of this in the built in default [`release_notes.md.j2`](https://github.com/python-semantic-release/python-semantic-release/blob/master/semantic_release/data/templates/release_notes.md.j2) file.

I happen to be using the exact same code for both release notes and changelog generation, so I need a solution that will work for both.

**Comment by maxrake** [comment]
@codejedi365, thank you for making the fix! PR #1005 has a code comment to "Remove in v10" and:

> It was never intended to be provided and will be removed in the future.

The [documentation for release notes context](https://python-semantic-release.readthedocs.io/en/latest/changelog_templates.html#release-notes-context) implies that `context` *should* be available for release notes:

> All of the changelog’s [template context](https://python-semantic-release.readthedocs.io/en/latest/changelog_templates.html#changelog-templates-template-rendering-template-context) is exposed to the [Jinja](https://jinja.palletsprojects.com/en/3.1.x/) template when rendering the release notes.

At least, that is how it was interpreted when adapting to the last major set of breaking changes.

If `context` is removed from the release notes rendering in a future release, can the elements folks commonly use in release notes be retained in another manner? For instance, can `prev_version` be exposed like `version` already is? Or, maybe a filter to provide the full/detailed changelog URL?

**Comment by codejedi365** [comment]
> The [documentation for release notes context](https://python-semantic-release.readthedocs.io/en/latest/changelog_templates.html#release-notes-context) implies that `context` *should* be available for release notes

Thanks for pointing that out. I didn't originally write the docs (ie didn't know that was there) and so maybe I was quick and incorrect to state it was not intended. When I refactored the code, it was not obvious which is why I unintentionally removed it.

Conceptually I don't think it makes sense to have the full history in a release notes but I fully realize there are a few specific desires. This is why I asked the questions above and leads to the answer to your next question:

> If `context` is removed from the release notes rendering in a future release, can the elements folks commonly use in release notes be retained in another manner? For instance, can `prev_version` be exposed like `version` already is? Or, maybe a filter to provide the full/detailed changelog URL?

Yes. My plan is to provide the common sensible ones automatically for you so that not everyone has to use jinja to extract the values.  Ones like `prev_release`, `prev_major_release`, `prev_minor_release`, etc. I have been working the past few days on more advanced and improved default release notes and changelogs for angular, which will be demoed & used by PSR itself. So anything I desire will be included as global variables in jinja.

Question: do we like using the context variable or is being a global fine? 

The problem with globals is that it imposes on users that want to use those keywords for their own macros or variable names. I'd rather not have a 4 chain of object attributes to get to the desired values but namespaces are helpful for preventing env pollution. I might add a variant of just `ctx` for more readable jinja.
